File: sainsmartrelay.py
#!/usr/bin/env python

import os, sys
import RPi.GPIO as gpio
import signal
import time
import copy
from threading import Timer

try:
    _TESTING_ = os.environ['TESTING']
except:
    _TESTING_ = False

PIN_NUMBERING_MODE = gpio.BCM

RelayPins = {
'PIN_RELAY_1'	: 17,
'PIN_RELAY_2'	: 27,
'PIN_RELAY_3'	: 22,
'PIN_RELAY_4'	: 10
}
RELAY_COUNT = len(RelayPins)

# Relay boolean constants. Flipped for Sainsmart relay module.
RELAY_ON = False;
RELAY_OFF = (not RELAY_ON);

# The default delay is 480 seconds (8 mins)
if _TESTING_:
    DEFAULT_DELAYSET = {'on_time':3, 'off_time':12, 'isset':False}
else:
    DEFAULT_DELAYSET = {'on_time':180, 'off_time':480, 'isset':False}

class Relay():
    def __init__(self):
        self.gpio_version = gpio.VERSION
        self.rpi_info = gpio.RPI_INFO
        print "GPIO VERSION: ", self.gpio_version
        print "RPI INFO: ", self.rpi_info
        gpio.setmode(PIN_NUMBERING_MODE);
        gpio.setup(RelayPins.values(), gpio.OUT)
        self.delayset = []
        for i in range(RELAY_COUNT):
            self.delayset.append(copy.copy(DEFAULT_DELAYSET))
        print "Relay setup done ", self.delayset

    def ALLON(self):
        print 'ALLON ...'
        gpio.output(RelayPins.values(), RELAY_ON)
        for i in range(self.device_count()):
            self.setOnDelay(i+1)

    def ALLOFF(self):
        print 'ALLOFF...'
        gpio.output(RelayPins.values(), RELAY_OFF)
        #for i in range(self.device_count()):
        #    self.setOffDelay(i+1)

    def ON_1(self):
        if self.isDelayed(1):
            return
        gpio.output(RelayPins['PIN_RELAY_1'], RELAY_ON);
        self.setOnDelay(1);
        print "Relay 1 on ", gpio.input(RelayPins['PIN_RELAY_1'])

    def ON_2(self):
        if self.isDelayed(2):
            return
        gpio.output(RelayPins['PIN_RELAY_2'], RELAY_ON);
        self.setOnDelay(2);
        print "Relay 2 on ", gpio.input(RelayPins['PIN_RELAY_2'])

    def ON_3(self):
        if self.isDelayed(3):
            return
        gpio.output(RelayPins['PIN_RELAY_3'], RELAY_ON);
        self.setOnDelay(3);
        print "Relay 3 on ", gpio.input(RelayPins['PIN_RELAY_3'])

    def ON_4(self):
        if self.isDelayed(4):
            return
        gpio.output(RelayPins['PIN_RELAY_4'], RELAY_ON);
        self.setOnDelay(4);
        print "Relay 4 on ", gpio.input(RelayPins['PIN_RELAY_4'])

    def OFF_1(self):
        if self.isDelayed(1):
            return
        gpio.output(RelayPins['PIN_RELAY_1'], RELAY_OFF);
        self.setOffDelay(1);
        print "Relay 1 off ", gpio.input(RelayPins['PIN_RELAY_1'])

    def OFF_2(self):
        if self.isDelayed(2):
            return
        gpio.output(RelayPins['PIN_RELAY_2'], RELAY_OFF);
        self.setOffDelay(2);
        print "Relay 2 off ", gpio.input(RelayPins['PIN_RELAY_2'])

    def OFF_3(self):
        if self.isDelayed(3):
            return
        gpio.output(RelayPins['PIN_RELAY_3'], RELAY_OFF);
        self.setOffDelay(3);
        print "Relay 3 off ", gpio.input(RelayPins['PIN_RELAY_3'])

    def OFF_4(self):
        if self.isDelayed(4):
            return
        gpio.output(RelayPins['PIN_RELAY_4'], RELAY_OFF);
        self.setOffDelay(3);
        print "Relay 4 off ", gpio.input(RelayPins['PIN_RELAY_4'])


    def device_count(self):
        return RELAY_COUNT

    def ON(self, id):
        if self.isDelayed(id):
            return
        gpio.output(RelayPins['PIN_RELAY_{}'.format(id)], RELAY_ON);
        self.setOnDelay(id)


    def OFF(self, id):
        if self.isDelayed(id):
            return
        gpio.output(RelayPins['PIN_RELAY_{}'.format(id)], RELAY_OFF);
        self.setOffDelay(id)

    def state(self):
        return list((gpio.input(RelayPins['PIN_RELAY_{}'.format(i+1)]),self.isDelayed(i+1)) for i in range(self.device_count()))

    def state_ORIG(self):
        mystate = []
        for dev in range(self.device_count()):
            this_dev = RelayPins['PIN_RELAY_{}'.format(dev+1)]
            mystate.append(gpio.input(this_dev))
        return mystate


    def isOn(self, id):
        if self.state()[id-1][0] == 0:
            #print '{} is ON'.format(id)
            return True
        else:
            #print '{} is OFF'.format(id)
            return False

    def isDelayed(self, id):
        return self.delayset[id-1]['isset']

    def setOnDelay(self, id):
        Timer(int(self.delayset[id-1]['on_time']), self.unsetDelay, [id]).start()
        self.delayset[id-1]['isset'] = True

    def setOffDelay(self, id):
        Timer(int(self.delayset[id-1]['off_time']), self.unsetDelay, [id]).start()
        self.delayset[id-1]['isset'] = True

    def unsetDelay(self, id):
        self.delayset[id-1]['isset'] = False
        #print "Relay %d delay now unset" % id

    def setDelaySetValue(self, id, key, val):
        self.delayset[id-1][key] = val

    def getDelaySetValue(self, id, key):
        return self.delayset[id-1][key]

if __name__=="__main__":
    relay = Relay()
    # Called on process interruption. Set all pins to "Input" default mode.
    def endProcess(signalnum = None, handler = None): 
        relay.ALLOFF()
        gpio.cleanup()
        sys.exit()

    signal.signal(signal.SIGINT, endProcess)

    while True:
        ct = raw_input("input: ")
        if ct == '1on':
            relay.ON_1()
        elif ct == '2on':
            relay.ON_2()
        elif ct == '3on':
            relay.ON_3()
        elif ct == '4on':
            relay.ON_4()
        elif ct == '1off':
            relay.OFF_1()
        elif ct == '2off':
            relay.OFF_2()
        elif ct == '3off':
            relay.OFF_3()
        elif ct == '4off':
            relay.OFF_4()
        elif ct == 'allon':
            relay.ALLON()
        elif ct == 'alloff':
            relay.ALLOFF()
        elif ct == 'state':
            relay.state()
        elif ct == '1isOn':
            relay.isOn(1)



# ex:set ai shiftwidth=4 inputtab=spaces smarttab noautotab:

File: src/scripts/modules/sainsmartrelay.js
//var rpio = require('rpio');
//import rpio from 'rpio';
import { default as rpio } from './jsogpio';


/* BCM numbering
var PossibleRelayPins = [
  18, 17, 27, 22,
  23, 24, 10, 9,
  11, 25, 8, 7,
  26, 16, 20, 21
];
var PIN_NUMBERING_MODE = 'gpio';
*/

/* Physical numbering */
var PossibleRelayPins = [
  12, 11, 13, 15,
  16, 18, 19, 21,
  22, 23, 32, 33,
  36, 37, 38, 40
];
var PIN_NUMBERING_MODE = 'physical';

//var RELAY_COUNT_MAX = PossibleRelayPins.length;

var RELAY_ON  = rpio.LOW;
var RELAY_OFF = rpio.HIGH;

var RelayPins = {};

var BREWTEST_DELAY_SET = {'on_time':3, 'off_time':12, 'isset':false};
var NORMAL_DELAY_SET = {'on_time':180, 'off_time':480, 'isset':false};
//var DEFAULT_DELAY_SET = NORMAL_DELAY_SET;

function Relay () {
  console.log("New Relay instance - available pins: " + PossibleRelayPins);

  this.brewtest = false;
  if (typeof process.env.BREWTEST !== 'undefined' && process.env.BREWTEST == 'true') {
    console.log("BREWTEST mode (" + process.env.BREWTEST + ")");
    this.brewtest = true;
  }

  var showConnectedPins = function () {
    var pins = [];
    for (var x in RelayPins) {pins.push(RelayPins[x]);}
    return pins;
  };

  var connectedRelays = function () {
    var counted = 0;

    for (var i=0;i<PossibleRelayPins.length;i++) {
      rpio.close(PossibleRelayPins[i]);
      //console.log("Pins: " + PossibleRelayPins[i]);
      rpio.open(PossibleRelayPins[i], rpio.OUTPUT, rpio.LOW);
    }
    for (i=0;i<PossibleRelayPins.length;i++) {
      rpio.mode(PossibleRelayPins[i], rpio.INPUT);
      //console.log("Pin %d = %d", PossibleRelayPins[i], rpio.read(PossibleRelayPins[i]));
      if (rpio.read(PossibleRelayPins[i]) == 1) {
        counted += 1;
        RelayPins['PIN_RELAY_' + counted] = PossibleRelayPins[i];
      }
    }
    // Reset for normal use - leave only connected pins open
    var connectedPins = showConnectedPins();
    for (i=0;i<PossibleRelayPins.length;i++) {
      rpio.close(PossibleRelayPins[i]);
      if (connectedPins.indexOf(PossibleRelayPins[i]) > -1) {
        rpio.open(PossibleRelayPins[i], rpio.OUTPUT, RELAY_OFF);
      }
    }
    return counted;
  };

  //rpio.init({gpiomem: true, mapping: PIN_NUMBERING_MODE, exportDelay: 8});
  rpio.init({gpiomem: true, mapping: PIN_NUMBERING_MODE});

  // Look for connected pins (populates RelayPins)
  this.relayCount = connectedRelays();

  // Delay sets
  this.delayset = [];
  for (var i=0;i<this.relayCount;i++) {
    this.delayset.push(JSON.parse(JSON.stringify(NORMAL_DELAY_SET)));
  }
  //console.log("DelaySets: " + JSON.stringify(this.delayset));

  console.log("Relay setup done. Using connected pins: " + showConnectedPins());
} 
export default Relay;

Relay.prototype.closeConnected = function () {
  var pins = [];
  for (var x in RelayPins) {pins.push(RelayPins[x]);}
  pins.forEach(function (item) {
    rpio.close(item);
  });
};

// Switch each connected pin on for 1 second
Relay.prototype.testConnected = function () {
  var pins = [];
  for (var x in RelayPins) {pins.push(RelayPins[x]);}
  pins.forEach(function (item) {
    //console.log("RRRR " + item)
    rpio.open(item, rpio.OUTPUT, RELAY_OFF);
    rpio.msleep(100);
    rpio.write(item, RELAY_ON);
    rpio.sleep(2);
    rpio.write(item, RELAY_OFF);
  });
};

Relay.prototype.deviceCount = function () {
  return this.relayCount;
};

Relay.prototype.isOn = function (id) {
  //if (this.state()[id-1][0] == RELAY_ON) {
  if (rpio.read(RelayPins['PIN_RELAY_'+id]) == RELAY_ON) {
    return true;
  } else {
    return false;
  }
};

Relay.prototype.isDelayed = function (id) {
  // Should first check if id-1 is valid!
  return this.delayset[id-1]['isset'];
};

Relay.prototype.ON = function (id) {
  if (this.isDelayed(id)) {
    return;
  }
  rpio.write(RelayPins['PIN_RELAY_'+id], RELAY_ON);
  this.setOnDelay(id);
};

Relay.prototype.OFF = function (id) {
  if (this.isDelayed(id)) {
    //console.log("Relay delayed");
    return;
  }
  rpio.write(RelayPins['PIN_RELAY_'+id], RELAY_OFF);
  this.setOffDelay(id);
};

Relay.prototype.unsetDelay = function (id) {
  this.delayset[id-1]['isset'] = false;
};

Relay.prototype.setOnDelay = function (id) {
  if (this.brewtest) {
    //console.log("setOnDelay() BREWTEST mode (" + BREWTEST_DELAY_SET['on_time'] + "secs)");
    setTimeout( function () { this.unsetDelay(id); }.bind(this), parseInt(BREWTEST_DELAY_SET['on_time'])*1000);
  } else {
    setTimeout( function () { this.unsetDelay(id); }.bind(this), parseInt(this.delayset[id-1]['on_time'])*1000);
  }
  this.delayset[id-1]['isset'] = true;
};

Relay.prototype.setOffDelay = function (id) {
  if (this.brewtest) {
    //console.log("setOffDelay() BREWTEST mode (" + BREWTEST_DELAY_SET['off_time'] + "secs)");
    setTimeout( function () { this.unsetDelay(id); }.bind(this), parseInt(BREWTEST_DELAY_SET['off_time'])*1000);
  } else {
    setTimeout( function () { this.unsetDelay(id); }.bind(this), parseInt(this.delayset[id-1]['off_time'])*1000);
  }
  this.delayset[id-1]['isset'] = true;
};

Relay.prototype.setDelaySetValue = function (id, key, val) {
  this.delayset[id-1][key] = val;
};

Relay.prototype.getDelaySetValue = function (id, key) {
  return this.delayset[id-1][key];
};

Relay.prototype.state = function () {
  var result = [];
  for (var i=0;i<this.relayCount;i++) {
    result.push([rpio.read(RelayPins['PIN_RELAY_'+parseInt(i+1)]), this.isDelayed(i+1)]);
  }
  return result;
};



/* ex:set ai shiftwidth=2 inputtab=spaces smarttab noautotab: */
